Compiler/test/inference.jl

```diff
@@ -3549,8 +3549,14 @@ f31974(n::Int) = f31974(1:n)
 # call cycles.
 @test code_typed(f31974, Tuple{Int}) !== nothing
 
-f_overly_abstract_complex() = Complex(Ref{Number}(1)[])
-@test Base.return_types(f_overly_abstract_complex, Tuple{}) == [Complex]
+# Issue #33472
+struct WrapperWithUnionall33472{T<:Real}
+    x::T
+end
+
+f_overly_abstract33472() = WrapperWithUnionall33472(Ref{Number}(1)[])
+# Check that this doesn't infer as `WrapperWithUnionall33472{T<:Number}`.
+@test Base.return_types(f_overly_abstract33472, Tuple{}) == [WrapperWithUnionall33472]
```

Comment:
This looks like it has a decent chance to be inferred as `WrapperWithUnionall33472{Int}`.

Reply:
Ok, I can write it as inferencebarrier with a type assert, which is the usual pattern. I was mostly just preserving what is already there.